### ISSUE TITLE
Issues with WORK1 in ?GEQR and ?GELQ

### ISSUE THREAD

**Issue by eugenechereshnev** [Reference-LAPACK/lapack#90]
I'm testing new QR routines and it seems that operations with WORK1 array won't be obvious for users. I see two possible issues here:

1) There is a convention in LAPACK that optimal size of workspace is returned in the first element of workspace but in the routines ?GEQR and ?GELQ it is returned in the second element for WORK1 and it can be misleading for users. It would be better to return optimal size of WORK1 in the first element, minimal size in the second element, the type of algorithm in the third etc (but only the optimal size may be returned - see the second issue).

2) Usually user passes one-element work array for workspace query. But in ?GEQR and ?GELQ its size should be at least 5 elements and it also may cause issues. So I think the better option here is to fill only the first element in workspace query - optimum size of WORK1 and fill other elements (row block size, column block size, algorithm type, ...) only inside computation call.

**Comment by langou** [comment]
Hi Eugene, 

I have voiced similar concerns with the proposed current interface. Let me voice them again. Let see what happens. Thanks for going in a direction that I support. 

I was wishing that the interface would be something like: 

(1) (WORK1, LWORK1) is workspace that needs to be kept for potential future usage (like applying the reflectors), (WORK2, LWORK2) is workspace that can be safely discarded after calling the DGEQR.

(2) For each workspace WORK, there is an LWORK, and workspace query is required. (We do not hardcode workspace sizes. If you want to know the workspace size, you have to perform a query.)

(3) LWORK=-1 is optimal workspace size query, we can use LWORK=-2 for minimal workspace size query. Queried workspace size returned in WORK1(1) and WORK2(1) respectively.

(4) We do not describe the internal data structure of WORK1(2:LWORK1). We just let the various LAPACK implementations should their own design. (For Netlib LAPACK, we will likely follow what is written.)

How does such an interface sound? 

This would answer both of your points, but this does more than just answering them, so I want to make clear that would be a good direction. Feel free to suggest an interface.

Best wishes, 
Julien.

**Comment by langou** [comment]
If we want to do minimal changes and address both of your points, (1) we can return optimal workspace in WORK1(1) (so first slot instead of second) and (2) we can only use WORK1(1) and not use WORK1(2:6) on the case when users do a workspace query. (As you suggest.) This is minimal changes.  One issue is that the only way for the user to obtain the minimal workspace is to execute the routine with the optimal workspace. We can imagine cases where it would be good to know the minimal workspace without having to allocate the optimal one first.

It would be good to know what interface you would want to see and everyone can pitch in their opinion even better. (Note for all: in case this is not clear, we are speaking of DGEQR.)

**Comment by eugenechereshnev** [comment]
Julien,

Your proposal is almost exactly what I was keeping in mind. You already wrote about its advantages. Besides that it will be easier to differentiate between ?GEQRF and ?GEQR interfaces. The first interface has well-defined format for reflectors. The second one provides internal format for possible optimizations (which is accessible through ?GEMQR calls). It will work well for us and for other vendors.

Now description of ?GEQR is not so clear at first sight:

```
*> DGEQR computes a QR factorization of an M-by-N matrix A,
*> using DLATSQR when A is tall and skinny
*> (M sufficiently greater than N), and otherwise DGEQRT:
*> A = Q * R .
```
I also had another concern about the name of WORK1. It fact it doesn't behave as work array. User can't destroy it after call and should pass it to consequent ?GEMQR call. But in the same time its size is not known in advance and must be obtained through query call - here it behaves as work array.

I found a similar case in LAPACK: routine ?STEMR has an argument Z to return eigenvectors and it supports size query because the number of eigenvectors is not known before call. Similarly, WORK1 can be renamed to something like TI (T **I**nternal) or VI (V **I**nternal) or something else and LWORK1 - to NTI or NVI with query support (and then WORK2 can be renamed to WORK). In that case the interface will be more consistent with other LAPACK.

**Comment by jwdemmel** [comment]
I have some extensive comments on our design goals for the interface to TSQR, as well as a response to the comments above. Please see
      http://people.eecs.berkeley.edu/~demmel/TSQR/TSQR_Interface_Design_Goals_v2.txt
I look forward to your comments.
Jim Demmel

**Comment by langou** [comment]
This was fixed by pull request #101. Thanks Konstantin and  Eugene and Intel. Julien.
